drivers/modem helper libraries refactoring

bjarki-andreasen opened the pull request:
**Introduction**
The drivers/modem contains libraries which are used for generic
functionality including uart interface, modem sockets and
command handling. These libraries work as expected, yet, they
lack documentation and APIs.

**Purpose**
This pull request adds well documented APIs to the libraries
and updates the in-tree drivers which depend on the libraries
to use the new APIs. The APIs are specifically designed to
prevent the need for directly accessing members of contexts,
which is the case for most of the initialization and use
of the libraries currently.

**Added feature (auto assigning socket id)**
The refactoring does not change the functionality except for
one added feature: A configuration option has been added to
the modem_socket library, which determines if the socket id
for a given socket is automatically assigned by the library,
which is the case for the quectel and simcom modems, or if
the id must be assigned later, as is the case with the ublox
sara modem.

This new feature comes with three helpers for determining if
the id has been assigned, if the socket has been allocated,
and for safely setting the id. With this new feature, the
simcom and quectel drivers have been updated to use the id,
which now behaves correctly with the added config option,
instead of the sock_fd.

**Status**
I have verified that all of the in-tree modem drivers can
compile with the new changes, and have tested and verified
the quectel-bg9x modem driver and the PPP GSM driver with
a quectel-bg95. @warasilapm has verified the ublox-sara-r4
driver.

The simcom7070 is nearly identical to the bg9x driver, so
the changes are expected to work for both of them, no one
with the modem has been available to test the update.

warasilapm commented on drivers/modem/modem_socket.h:
```diff
@@ -82,7 +86,65 @@ int modem_socket_poll_prepare(struct modem_socket_config *cfg, struct modem_sock
 void modem_socket_data_ready(struct modem_socket_config *cfg, struct modem_socket *sock);
-int modem_socket_init(struct modem_socket_config *cfg, const struct socket_op_vtable *vtable);
+
+/**
```
`one_based_id` is strange to read and does not communicate the purpose as much as `base_socket_num`. I also don't see any reason this can't be something other than zero or one if a modem manufacturer does something particularly insane, so I'd leave this as an `int` to prevent a future refactor.

bjarki-trackunit replied:
We need to change the behavior a bit more if 0 and 1 are not the only options, the id is set to a reserved value which is defined as socket_len + 1, as this value is always one higher than the max index (if base_socket_num is 1 and socket_len is 5, max socket id is 5, yet socket_len + 1 is 6). The value indicates that the socket has been allocated, but has not been assigned an id. if we would need to set the reserved value to **socket_len + base_socket_num** to acheive this, which we can easily do

warasilapm replied:
That's something I stashed at some point, since it's kind of silly to just add one instead of adding the base socket number. Further, I think the whole socket number system needs to be a set of macros like:
``` C
#define MODEM_SOCKET_ID_UNUSED(CFG) ((CFG).base_socket_num - 1)
#define MODEM_SOCKET_ID_NEW(CFG) ((CFG).base_socket_num - 1)
#define MODEM_SOCKET_ID_IS_VALID(CFG, ID) (((CFG).base_socket_num <= (ID)) && (((CFG).base_socket_num + (CFG).sockets_len) > (ID)))
```
I just haven't had time to do this refactor and hunt down all comparisons of `socket_id` which uses these.

warasilapm replied:
I still think we need a clear documentation in the API for what range of values for a socket id are valid as well as what the special values for "unallocated" and "unassigned" are. I would have added a comment new the `id` member of `struct modem_socket` but github doesn't empower me with that option for the scope of your PR.

ycsin commented:
cc @bbilas

ycsin commented on drivers/modem/modem_cmd_handler.c:
```diff
@@ -645,28 +645,70 @@ void modem_cmd_handler_tx_unlock(struct modem_cmd_handler *handler)
 {
-	if (!handler || !data) {
+	/* Verify arguments */
+	if (!handler || !data || !match_buf || !match_buf_len || !buf_pool) {
```
probably better to compare them against NULL or number to be MISRA-compliant

ycsin commented on drivers/modem/modem_cmd_handler.c:
```diff
@@ -645,28 +645,70 @@ void modem_cmd_handler_tx_unlock(struct modem_cmd_handler *handler)
+	data->alloc_timeout = alloc_timeout;
+
+	/* Process end of line */
+	if (!eol) {
```
compare with `NULL`

ycsin commented on drivers/modem/modem_cmd_handler.c:
```diff
@@ -645,28 +645,70 @@ void modem_cmd_handler_tx_unlock(struct modem_cmd_handler *handler)
+
+	/* Process end of line */
+	if (!eol) {
+		data->eol = NULL;
```
looks like you can just set `data->eol` to `eol` outside of this if, and set the len accordingly with this if statement

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -174,10 +174,11 @@ static void gsm_rx(struct gsm_modem *gsm)
-		/* The handler will listen AT channel */
-		gsm->context.cmd_handler.process(&gsm->context.cmd_handler,
+		/* Process data */
+		modem_cmd_handler_process(&gsm->context.cmd_handler,
```
since the args arent aligned anymore, probably can just make it into a single line

ycsin reviewed (commented):
Looks like a major cleanup, hopefully I have time for the meat of the PR during weekend, thanks for this!

warasilapm commented:
Since @ycsin brought up MISRA compliance, the ublox-sara-r4 file could probably use a once over by someone who has better ingested the new(ish) coding guidelines published in the docs. I'll be opening a PR for sara-r4 fixes in the near future and it would be good to get a feel for what following those guidelines entails and a good opportunity to make the changes while I'm touching most of the files.

bjarki-trackunit commented:
> Since @ycsin brought up MISRA compliance, the ublox-sara-r4 file could probably use a once over by someone who has better ingested the new(ish) coding guidelines published in the docs. I'll be opening a PR for sara-r4 fixes in the near future and it would be good to get a feel for what following those guidelines entails and a good opportunity to make the changes while I'm touching a bunch of the file.
> 
> As an aside, does anyone know what the plan is for getting net_buf et.al. compliant with the coding guidelines is? I looked around on the mailing list for stuff relating to that but didn't find anything.

Regarding the ublox-sara-r4, the order op operations would preferably be:
1. Test this PR to make sure ublox-sara-r4 is working (which i can't since i don't have that modem)
2. Merge this PR
3. Start work on fixes for ublox-sara-r4

I know nothing regarding net_buf

warasilapm commented:
> > Since @ycsin brought up MISRA compliance, the ublox-sara-r4 file could probably use a once over by someone who has better ingested the new(ish) coding guidelines published in the docs. I'll be opening a PR for sara-r4 fixes in the near future and it would be good to get a feel for what following those guidelines entails and a good opportunity to make the changes while I'm touching a bunch of the file.
> > As an aside, does anyone know what the plan is for getting net_buf et.al. compliant with the coding guidelines is? I looked around on the mailing list for stuff relating to that but didn't find anything.
> 
> Regarding the ublox-sara-r4, the order op operations would preferably be:
> 
> 1. Test this PR to make sure ublox-sara-r4 is working (which i can't since i don't have that modem)
> 2. Merge this PR
> 3. Start work on fixes for ublox-sara-r4
> 
> I know nothing regarding net_buf

The only issue with this approach is that parts of the SARA-R4 driver were broken before. I can just rebase my fixes on that branch, but the SARA being "working" after this PR is a bit of a misnomer without further fixes that shouldn't be wrapped up in this single refactor commit.

bjarki-trackunit commented:
> > > Since @ycsin brought up MISRA compliance, the ublox-sara-r4 file could probably use a once over by someone who has better ingested the new(ish) coding guidelines published in the docs. I'll be opening a PR for sara-r4 fixes in the near future and it would be good to get a feel for what following those guidelines entails and a good opportunity to make the changes while I'm touching a bunch of the file.
> > > As an aside, does anyone know what the plan is for getting net_buf et.al. compliant with the coding guidelines is? I looked around on the mailing list for stuff relating to that but didn't find anything.
> > 
> > 
> > Regarding the ublox-sara-r4, the order op operations would preferably be:
> > 
> > 1. Test this PR to make sure ublox-sara-r4 is working (which i can't since i don't have that modem)
> > 2. Merge this PR
> > 3. Start work on fixes for ublox-sara-r4
> > 
> > I know nothing regarding net_buf
> 
> The only issue with this approach is that parts of the SARA-R4 driver were broken before. I can just rebase my fixes on this branch, but the SARA being "working" after this PR is a bit of a misnomer without further fixes that shouldn't be wrapped up in this single refactor commit.

I see, so order of operations is
1. Merge this PR with main
2. Rebase SARA fixes on this PR (or main)
3. Merge SARA fixes with main

Or
1. Rebase SARA fixes on this PR
2. Merge SARA fixes to this PR
3. Merge this PR with main

If i understand correctly

warasilapm commented:
> > > > Since @ycsin brought up MISRA compliance, the ublox-sara-r4 file could probably use a once over by someone who has better ingested the new(ish) coding guidelines published in the docs. I'll be opening a PR for sara-r4 fixes in the near future and it would be good to get a feel for what following those guidelines entails and a good opportunity to make the changes while I'm touching a bunch of the file.
> > > > As an aside, does anyone know what the plan is for getting net_buf et.al. compliant with the coding guidelines is? I looked around on the mailing list for stuff relating to that but didn't find anything.
> > > 
> > > 
> > > Regarding the ublox-sara-r4, the order op operations would preferably be:
> > > 
> > > 1. Test this PR to make sure ublox-sara-r4 is working (which i can't since i don't have that modem)
> > > 2. Merge this PR
> > > 3. Start work on fixes for ublox-sara-r4
> > > 
> > > I know nothing regarding net_buf
> > 
> > 
> > The only issue with this approach is that parts of the SARA-R4 driver were broken before. I can just rebase my fixes on this branch, but the SARA being "working" after this PR is a bit of a misnomer without further fixes that shouldn't be wrapped up in this single refactor commit.
> 
> I see, so order of operations is
> 
> 1. Merge this PR with main
> 2. Rebase SARA fixes on this PR (or main)
> 3. Merge SARA fixes with main
> 
> Or
> 
> 1. Rebase SARA fixes on this PR
> 2. Merge SARA fixes to this PR
> 3. Merge this PR with main
> 
> If i understand correctly, and you prefer the first one

That's correct. It's not necessarily a problem to have them become part of this PR, but they shouldn't be squashed in with the existing refactor commit and it expands the scope of this PR so it may take longer to get this one merged in.

bjarki-trackunit commented:
> > > > > Since @ycsin brought up MISRA compliance, the ublox-sara-r4 file could probably use a once over by someone who has better ingested the new(ish) coding guidelines published in the docs. I'll be opening a PR for sara-r4 fixes in the near future and it would be good to get a feel for what following those guidelines entails and a good opportunity to make the changes while I'm touching a bunch of the file.
> > > > > As an aside, does anyone know what the plan is for getting net_buf et.al. compliant with the coding guidelines is? I looked around on the mailing list for stuff relating to that but didn't find anything.
> > > > 
> > > > 
> > > > Regarding the ublox-sara-r4, the order op operations would preferably be:
> > > > 
> > > > 1. Test this PR to make sure ublox-sara-r4 is working (which i can't since i don't have that modem)
> > > > 2. Merge this PR
> > > > 3. Start work on fixes for ublox-sara-r4
> > > > 
> > > > I know nothing regarding net_buf
> > > 
> > > 
> > > The only issue with this approach is that parts of the SARA-R4 driver were broken before. I can just rebase my fixes on this branch, but the SARA being "working" after this PR is a bit of a misnomer without further fixes that shouldn't be wrapped up in this single refactor commit.
> > 
> > 
> > I see, so order of operations is
> > 
> > 1. Merge this PR with main
> > 2. Rebase SARA fixes on this PR (or main)
> > 3. Merge SARA fixes with main
> > 
> > Or
> > 
> > 1. Rebase SARA fixes on this PR
> > 2. Merge SARA fixes to this PR
> > 3. Merge this PR with main
> > 
> > If i understand correctly, and you prefer the first one
> 
> That's correct. It's not necessarily a problem to have them become part of this PR, but they shouldn't be squashed in with the existing refactor commit and it expands the scope of this PR so it may take longer to get this one merged in.

If you can confirm that the SARA modem driver is at least functional once this PR is merged, without the SARA fix PR, then we can go ahead and merge this PR so we can proceed with the SARA fix PR. I will test the ppp and bg9x driver to make sure they work as well, the simcom is a copy of the bg9x (or the other way around) so if i match the changes in the bg9x driver with the simcom, we should be good to go.

warasilapm commented:
I have not had time to diagnose the issue, but the sara driver is at least partially broken with the current state of this PR.

bjarki-trackunit commented:
> I have not had time to diagnose the issue, but the sara driver is at least partially broken with the current state of this PR. I tested with the fixes required from my fork and UDP sockets are non-functional.

I can verify the BG9x and PPP drivers are working. The wifi/esp_at.c has had very minor changes to it, and the simcom7080 has had nearly identical changes made to it as the BG9x, since they are nearly identical drivers, I have high certainty that these drivers will work.

Once you have had time to diagnose the issue with the SARA-R4 changes in this PR, and it has been fixed, I will turn this from a draft to a PR so we can have it merged :)

bjarki-trackunit commented:
Latest update to status:
- **quectel-bg9x** tested and working
- **ublox-sara-r4** tested and working
- **gsm_ppp** tested and working
- **simcom-sim7080** builds but not tested, changes are identical to quectel-bg9x, so expected to work
- **esp.c** builds but not tested, only initialization of modem_cmd_handler and modem_iface_uart updated

The fix is now in a state where it can be merged with confidence.

warasilapm commented on drivers/modem/modem_iface_uart.h:
```diff
@@ -55,17 +51,35 @@ int modem_iface_uart_init_dev(struct modem_iface *iface,
-			  const struct device *dev);
+static inline int modem_iface_uart_rx_wait(struct modem_iface *iface, k_timeout_t timeout)
+{
+	return k_sem_take(&((struct modem_iface_uart_data *)(iface->iface_data))->rx_sem, timeout);
```
The cast that happens here is ugly but I understand why it has to occur. While the UART interface is the only one that exists today, it seems to me that this sort of rx semaphore belongs in `struct modem_iface` rather than as an implementation detail for UART specifically. Not really sure if this is worth changing at the moment, but might be an abstraction worth adding to the modem context header for all potential interfaces going forwards. Then again, it may never really matter.

warasilapm replied:
If I take my reasoning one step further, should the rx thread which waits on the rx semaphore, processes the data, yields, and then loops be a standard part of the modem context API?

warasilapm commented on drivers/modem/modem_socket.c:
```diff
@@ -236,7 +237,7 @@ void modem_socket_put(struct modem_socket_config *cfg, int sock_fd)
 	k_sem_take(&cfg->sem_lock, K_FOREVER);
 
-	sock->id = cfg->base_socket_num - 1;
+	sock->id = -1;
```
I know this is going to sound silly, but I think that in all cases the "not allocated" value should be `cfg->base_socket_num - 1` rather than `-1` outright. I know it seems ridiculous to suggest that a manufacturer would ever use a socket id of negative numbers, but subtracting 1 always guarantees it's out of the range of valid socket ids.

warasilapm commented on drivers/modem/modem_cmd_handler.h:
```diff
@@ -266,15 +266,54 @@ int modem_cmd_handler_setup_cmds_nolock(struct modem_iface *iface,
+			      k_timeout_t alloc_timeout, const char *eol, void *user_data);
+
+/**
+ * @brief Set response and unsolicitet command handlers
```
Small typo: unsolicitet should be unsolicited.

warasilapm commented on drivers/modem/modem_cmd_handler.h:
```diff
@@ -266,15 +266,54 @@ int modem_cmd_handler_setup_cmds_nolock(struct modem_iface *iface,
-			   struct modem_cmd_handler_data *data);
+int modem_cmd_handler_init_cmds(struct modem_cmd_handler *handler,
+			      const struct modem_cmd *response_cmds, size_t response_cmds_len,
+			      const struct modem_cmd *unsol_cmds, size_t unsol_cmds_len);
```
While in general I appreciate the new modem initialization call signature pattern, I don't entirely understand the reasoning for separating out the command handler commands initialization into a separate call from the previous initialization call.

warasilapm reviewed (dismissed):
While there are still things to improve in the modem context API, this is definitely a good step forward for the API. Nothing I commented on is pointing out something which is broken, and this PR could be merged as is and the API iterated on later.

One thing I did want to add is that this PR consists of one, relatively large commit. I would have liked to see the changes to the APIs broken up into smaller more focused commits to make review and diagnosis of the previous ublox-sara-r4 driver bug easier. This is part of the commit guidelines for Zephyr and was emphasized by the talk given at ZDS2022 regarding contribution. For example, the refactor of `modem_uart_iface.h` and its associated source should be it's own self contained commit, then the `modem_socket.h` changes, and so on. There's also some white space and documentation changes that would have been easier to discard if committed separately. Maybe I'm nitpicking, but this is a bear of a commit to grok, especially given my limited previous understanding of the finer points of the modem context code and the poor existing documentation on how the modem context and its constituents play together.

warasilapm commented:
@rerickson1 Now that we've passed the 3.2 release, is this something we can get merged in? I think it's ready to go.

rerickson1 commented:
> @rerickson1 Now that we've passed the 3.2 release, is this something we can get merged in? I think it's ready to go.

Just need two reviews on this. I have never worked with this code so I welcome other reviewers on this to approve.

mniestroj commented on drivers/modem/gsm_ppp.c:
```diff
@@ -1267,24 +1265,24 @@ static int gsm_init(const struct device *dev)
-				   &gsm->cmd_handler_data);
+	ret = modem_cmd_handler_init(&gsm->context.cmd_handler, &gsm->cmd_handler_data,
+		&gsm->cmd_match_buf[0], sizeof(gsm->cmd_match_buf), &gsm_recv_pool,
+		K_NO_WAIT, "\r", NULL);
```
I don't think calling `modem_cmd_handler_init()` is readable enough now. IMO there are way too many arguments passed to this function. This makes it also error prone, as it is easy to pass arguments in wrong order, or pass something unrelevant which is just the same type.

What about defining a structure (possibly const to save RAM space), that would be initialized using "designated initializer", similar how work queue is initialized with help of `struct k_work_queue_config`?

warasilapm replied:
It's interesting that this approach is suggested considering that's how it was done previously. I think the reasoning for moving them to function arguments was to facilitate runtime checking of the parameters before they were written into the context structures.

mniestroj reviewed (commented):
Thanks for this PR!

I didn't go through all changes, as it seems they could be split one commit for one logical change (e.g. API change / introduction) for easier review.

bjarki-trackunit commented:
The following updates to this PR have been performed based on requests:

**Changes**
* PR has been split into individual commits pr feature
* modem_cmd_handler_init() has been updated to use a "designated initializer" as suggested by @mniestroj
* modem_cmd_handler_init_cmds() has been merged with modem_cmd_handler_init() and removed, there is not only one init function for modem_cmd_handler.
* Drivers have been updated to use the new modem_cmd_handler_init()
* Whitespace changes from gsm_ppp.c have been removed.

**Validation**
gsm_ppp and quectel_bg9x have been tested with the new modem_cmd_handler_init(). All modem drivers can compile. Testing ublox-sara and esp_at drivers is encouraged.

warasilapm commented on drivers/modem/modem_cmd_handler.c:
```diff
@@ -646,27 +646,46 @@ void modem_cmd_handler_tx_unlock(struct modem_cmd_handler *handler)
 }
 
 int modem_cmd_handler_init(struct modem_cmd_handler *handler,
-			   struct modem_cmd_handler_data *data)
+			      struct modem_cmd_handler_data *data,
+			      const struct modem_cmd_handler_setup *setup)
```
This has some strange whitespace issue.

warasilapm commented on drivers/modem/modem_cmd_handler.h:
```diff
@@ -299,6 +338,31 @@ int modem_cmd_handler_tx_lock(struct modem_cmd_handler *handler,
+ * @param iface The interface which receives incoming data
+ */
+static inline void modem_cmd_handler_process(struct modem_cmd_handler *handler,
+			      struct modem_iface *iface)
```
Another strange whitespace situation.

warasilapm commented on drivers/modem/modem_iface_uart.h:
```diff
@@ -55,17 +51,35 @@ int modem_iface_uart_init_dev(struct modem_iface *iface,
+ * @return -EINVAL if any argument is invalid
+ * @return 0 if successful
+ */
+int modem_iface_uart_init(struct modem_iface *iface, struct modem_iface_uart_data *data,
+			      char *rx_rb_buf, size_t rx_rb_buf_len, const struct device *dev,
+			      bool hw_flow_control);
```
And a third whitespace strangeness.

warasilapm commented on drivers/modem/modem_iface_uart_async.c:
```diff
@@ -148,9 +148,9 @@ int modem_iface_uart_init_dev(struct modem_iface *iface,
-int modem_iface_uart_init(struct modem_iface *iface,
-			  struct modem_iface_uart_data *data,
-			  const struct device *dev)
+int modem_iface_uart_init(struct modem_iface *iface, struct modem_iface_uart_data *data,
+			char *rx_rb_buf, size_t rx_rb_buf_len, const struct device *dev,
+			bool hw_flow_control)
```
A fourth whitespace strangeness.

warasilapm commented on drivers/modem/modem_iface_uart_interrupt.c:
```diff
@@ -198,9 +198,9 @@ int modem_iface_uart_init_dev(struct modem_iface *iface,
-int modem_iface_uart_init(struct modem_iface *iface,
-			  struct modem_iface_uart_data *data,
-			  const struct device *dev)
+int modem_iface_uart_init(struct modem_iface *iface, struct modem_iface_uart_data *data,
+			char *rx_rb_buf, size_t rx_rb_buf_len, const struct device *dev,
+			bool hw_flow_control)
```
This is the fifth whitespace inconsistency I've found. I'm going to stop pointing them out at this point.

warasilapm reviewed (commented):
Tested the current state of the PR on the ublox-sara-r4 and was successful.

There is some weird non-aligned whitespace that either got added or I didn't notice before. It's inconsistent between different function definitions. Maybe your tab size isn't set to 8 or something? Can't be sure. It doesn't bother me all that much but it would be better if it were kept aligned nicely as before since this is Zephyr code.

bjarki-trackunit commented:
@mniestroj Can you find time to re-review this PR? The following comment provides the updated status since you last requested changes :) https://github.com/zephyrproject-rtos/zephyr/pull/48530#issuecomment-1286774121

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -1267,19 +1267,26 @@ static int gsm_init(const struct device *dev)
+		.response_cmds = response_cmds,
+		.response_cmds_len = ARRAY_SIZE(response_cmds),
+		.unsol_cmds = NULL,
+		.unsol_cmds_len = 0
```
```suggestion
		.unsol_cmds_len = 0,
```

ycsin commented on drivers/modem/modem_cmd_handler.c:
```diff
@@ -646,27 +645,46 @@ void modem_cmd_handler_tx_unlock(struct modem_cmd_handler *handler)
-	if (!data->match_buf_len) {
+	/* Verify setup */
+	if (setup->match_buf == NULL || setup->match_buf_len == 0 || setup->buf_pool == NULL
+		|| (NULL != setup->response_cmds && 0 == setup->response_cmds_len)
```
Normally the operator should be at the right?

```suggestion
		(NULL != setup->response_cmds && 0 == setup->response_cmds_len) ||
```

and not sure why we changed to yoda conditions from this line?

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -1299,13 +1306,10 @@ static int gsm_init(const struct device *dev)
 
-	ret = modem_iface_uart_init(&gsm->context.iface, &gsm->gsm_data,
-				DEVICE_DT_GET(GSM_UART_NODE));
+	ret = modem_iface_uart_init(&gsm->context.iface, &gsm->gsm_data, &gsm->gsm_rx_rb_buf[0],
```
same as above, the understandability decreased, but I'm not sure if the incentive is high enough to group the rb_buf & uart related into their own group of structs

ycsin commented on drivers/modem/modem_cmd_handler.c:
```diff
@@ -646,27 +645,46 @@ void modem_cmd_handler_tx_unlock(struct modem_cmd_handler *handler)
+	/* Initialize command handler data members */
 	k_sem_init(&data->sem_tx_lock, 1, 1);
 	k_sem_init(&data->sem_parse_lock, 1, 1);
-
```
IMO I'd like to preserve this empty line

ycsin commented on drivers/modem/modem_context.h:
```diff
@@ -80,7 +80,9 @@ struct modem_context {
  * @retval 0 if ok, < 0 if error.
  */
-int modem_context_sprint_ip_addr(const struct sockaddr *addr, char *buf, size_t buf_size);
+int modem_context_sprint_ip_addr(const struct sockaddr *addr,
+				 char *buf,
+				 size_t buf_size);
```
ditto

ycsin commented on drivers/modem/modem_socket.c:
```diff
@@ -173,9 +173,10 @@ int modem_socket_get(struct modem_socket_config *cfg, int family, int type, int
+		(cfg->base_socket_id + cfg->sockets_len);
 	z_finalize_fd(cfg->sockets[i].sock_fd, &cfg->sockets[i],
-		      (const struct fd_op_vtable *)cfg->vtable);
+		(const struct fd_op_vtable *)cfg->vtable);
```
is this the result of the `.clang-format`? previously it was aligned

ycsin commented on drivers/modem/modem_socket.c:
```diff
@@ -410,18 +411,78 @@ void modem_socket_data_ready(struct modem_socket_config *cfg, struct modem_socke
 
-int modem_socket_init(struct modem_socket_config *cfg, const struct socket_op_vtable *vtable)
+int modem_socket_init(struct modem_socket_config *cfg, struct modem_socket *sockets,
+			      size_t sockets_len, int base_socket_id, bool assign_id,
+			      const struct socket_op_vtable *vtable)
```
not aligned?

ycsin commented on drivers/modem/modem_socket.c:
```diff
@@ -410,18 +411,78 @@ void modem_socket_data_ready(struct modem_socket_config *cfg, struct modem_socke
+{
+	/* Verify socket is assigned to a valid value */
+	if (cfg->base_socket_id <= sock->id
+		&& sock->id < (cfg->base_socket_id + cfg->sockets_len)) {
```
&& should be at the right?

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -1299,13 +1306,10 @@ static int gsm_init(const struct device *dev)
-	ret = modem_iface_uart_init(&gsm->context.iface, &gsm->gsm_data,
-				DEVICE_DT_GET(GSM_UART_NODE));
+	ret = modem_iface_uart_init(&gsm->context.iface, &gsm->gsm_data, &gsm->gsm_rx_rb_buf[0],
+		sizeof(gsm->gsm_rx_rb_buf), DEVICE_DT_GET(GSM_UART_NODE),
+		DT_PROP(GSM_UART_NODE, hw_flow_control));
```
should be aligned

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -1267,19 +1267,26 @@ static int gsm_init(const struct device *dev)
+
+	/* cmd handler */
+	ret = modem_cmd_handler_init(&gsm->context.cmd_handler, &gsm->cmd_handler_data,
+		&cmd_handler_setup);
```
should be aligned?

ycsin reviewed (dismissed):
apart from the comments, generally its a +1 from me

ycsin commented on drivers/modem/modem_iface_uart.h:
```diff
@@ -55,17 +51,36 @@ int modem_iface_uart_init_dev(struct modem_iface *iface,
-int modem_iface_uart_init(struct modem_iface *iface,
-			  struct modem_iface_uart_data *data,
-			  const struct device *dev);
+static inline int modem_iface_uart_rx_wait(struct modem_iface *iface,
```
can fit into a single line

ycsin commented on drivers/modem/modem_socket.c:
```diff
@@ -410,18 +411,78 @@ void modem_socket_data_ready(struct modem_socket_config *cfg, struct modem_socke
 
-	cfg->vtable = vtable;
+bool modem_socket_is_allocated(const struct modem_socket_config *cfg,
+			      const struct modem_socket *sock)
```
alignment

ycsin commented on drivers/modem/modem_socket.c:
```diff
@@ -410,18 +411,78 @@ void modem_socket_data_ready(struct modem_socket_config *cfg, struct modem_socke
+}
+
+int modem_socket_id_assign(const struct modem_socket_config *cfg,
+			      struct modem_socket *sock, int id)
```
alignment

ycsin commented on drivers/modem/modem_socket.h:
```diff
@@ -82,7 +86,66 @@ int modem_socket_poll_prepare(struct modem_socket_config *cfg, struct modem_sock
+ * @return true if the socket has been allocated, otherwise false
+ */
+bool modem_socket_is_allocated(const struct modem_socket_config *cfg,
+			      const struct modem_socket *sock);
```
alignment

ycsin replied:
some of the remaining ones from last review, ping @bjarki-trackunit

ycsin commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -734,7 +735,7 @@ static int offload_connect(void *obj, const struct sockaddr *addr,
 	/* Formulate the complete string. */
-	snprintk(buf, sizeof(buf), "AT+QIOPEN=%d,%d,\"%s\",\"%s\",%d,0,0", 1, sock->sock_fd, protocol,
+	snprintk(buf, sizeof(buf), "AT+QIOPEN=%d,%d,\"%s\",\"%s\",%d,0,0", 1, sock->id, protocol,
 		ip_str, dst_port);
```
not in scope but should be aligned

ycsin commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -1145,34 +1147,36 @@ static int modem_init(const struct device *dev)
-	mdata.socket_config.base_socket_num = MDM_BASE_SOCKET_NUM;
-	ret = modem_socket_init(&mdata.socket_config, &offload_socket_fd_op_vtable);
+	ret = modem_socket_init(&mdata.socket_config, &mdata.sockets[0], ARRAY_SIZE(mdata.sockets),
+			   MDM_BASE_SOCKET_NUM, true, &offload_socket_fd_op_vtable);
```
alignment

ycsin commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -1145,34 +1147,36 @@ static int modem_init(const struct device *dev)
-	mdata.cmd_handler_data.eol		   = "\r\n";
-	ret = modem_cmd_handler_init(&mctx.cmd_handler, &mdata.cmd_handler_data);
+	ret = modem_cmd_handler_init(&mctx.cmd_handler, &mdata.cmd_handler_data,
+		&cmd_handler_setup);
```
alignment

ycsin commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -1145,34 +1147,36 @@ static int modem_init(const struct device *dev)
-	ret = modem_iface_uart_init(&mctx.iface, &mdata.iface_data,
-				    MDM_UART_DEV);
+	ret = modem_iface_uart_init(&mctx.iface, &mdata.iface_data, &mdata.iface_rb_buf[0],
+		sizeof(mdata.iface_rb_buf), MDM_UART_DEV, DT_PROP(MDM_UART_NODE, hw_flow_control));
```
alignment

ycsin commented on drivers/modem/simcom-sim7080.c:
```diff
@@ -2337,35 +2338,38 @@ static int modem_init(const struct device *dev)
+
+	/* cmd handler */
+	ret = modem_cmd_handler_init(&mctx.cmd_handler, &mdata.cmd_handler_data,
+		&cmd_handler_setup);
```
alignment

ycsin commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -737,15 +737,22 @@ static const struct setup_cmd query_cellinfo_cmds[] = {
 	if (sock) {
-		sock->id = ATOI(argv[0],
-				mdata.socket_config.base_socket_num - 1,
+		id = ATOI(argv[0],
+				-1,
 				"socket_id");
```
probably can make these into a single line

ycsin replied:
some of the remaining ones from last review, ping @bjarki-trackunit

ycsin commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -2138,38 +2147,36 @@ static int modem_init(const struct device *dev)
-	ret = modem_cmd_handler_init(&mctx.cmd_handler,
-				     &mdata.cmd_handler_data);
+	ret = modem_cmd_handler_init(&mctx.cmd_handler, &mdata.cmd_handler_data,
+		&cmd_handler_setup);
```
alignment

ycsin commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -2138,38 +2147,36 @@ static int modem_init(const struct device *dev)
-	ret = modem_iface_uart_init(&mctx.iface, &mdata.iface_data,
-				    MDM_UART_DEV);
+	ret = modem_iface_uart_init(&mctx.iface, &mdata.iface_data, &mdata.iface_rb_buf[0],
+		sizeof(mdata.iface_rb_buf), MDM_UART_DEV, DT_PROP(MDM_UART_NODE, hw_flow_control));
```
alignment

ycsin commented on drivers/wifi/esp_at/esp.c:
```diff
@@ -1138,28 +1138,31 @@ static int esp_init(const struct device *dev)
-	ret = modem_cmd_handler_init(&data->mctx.cmd_handler,
-				       &data->cmd_handler_data);
+	ret = modem_cmd_handler_init(&data->mctx.cmd_handler, &data->cmd_handler_data,
+		&cmd_handler_setup);
```
alignment

ycsin commented on drivers/wifi/esp_at/esp.c:
```diff
@@ -1138,28 +1138,31 @@ static int esp_init(const struct device *dev)
-	ret = modem_iface_uart_init(&data->mctx.iface, &data->iface_data,
-				    DEVICE_DT_GET(DT_INST_BUS(0)));
+	ret = modem_iface_uart_init(&data->mctx.iface, &data->iface_data, &data->iface_rb_buf[0],
+		sizeof(data->iface_rb_buf), DEVICE_DT_GET(DT_INST_BUS(0)),
+		DT_PROP(ESP_BUS, hw_flow_control));
```
alignment

bjarki-trackunit commented:
@ycsin Adressed all formatting issues you pointed out, please re review :)

ycsin commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -2138,38 +2147,37 @@ static int modem_init(const struct device *dev)
+		.response_cmds = response_cmds,
+		.response_cmds_len = ARRAY_SIZE(response_cmds),
+		.unsol_cmds = unsol_cmds,
+		.unsol_cmds_len = ARRAY_SIZE(unsol_cmds)
```
ditto

ycsin commented on drivers/wifi/esp_at/esp.c:
```diff
@@ -1139,28 +1139,31 @@ static int esp_init(const struct device *dev)
+		.response_cmds = response_cmds,
+		.response_cmds_len = ARRAY_SIZE(response_cmds),
+		.unsol_cmds = unsol_cmds,
+		.unsol_cmds_len = ARRAY_SIZE(unsol_cmds)
```
ditto

ycsin commented on drivers/modem/modem_cmd_handler.c:
```diff
@@ -646,25 +645,47 @@ void modem_cmd_handler_tx_unlock(struct modem_cmd_handler *handler)
-	if (!data->match_buf_len) {
+	/* Verify setup */
+	if ((setup->match_buf == NULL) ||
+		(setup->match_buf_len == 0) ||
```
this looks abit weird, I wonder if this is the result of our clang-format?

mniestroj reviewed (dismissed):
Please provide motivation in every commit (https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-body) about why each API change is introduced (see previous comments).

Currently this patchset is not bisectable (API changes are separate from driver updates). And there are still style (alignment/indentation) issues. Please also squash the last commit, as it only fixes previous commits with style improvements.

Commit:
drivers/modem/modem_iface_uart: Update API

The UART IFACE API currently exposes the context struct
modem_iface_uart_data, expecting the user to fill in some
of the fields, with no documentation specifying which fields
and what they mean.

This API update moves all user configurable values in the context
out into a config struct, modem_iface_uart_config, within which
members are documented.

This prevents the user from interacting directly with the context
making use of the library safer and more readable.

The config structure helps make code readable by using "named args"
in a const struct instead of a long, nameless, parameter list passed
to the modem_iface_uart_init function.

The new API function modem_iface_uart_rx_wait is added to prevent the
user from having to interact with the rx sem in the context directly.

The context can now be safely interated with without direct access to
any of its members.

Pointers to the ring buffer params in the context have been moved to
config struct, as these are only useful during initialization of the
context.

Signed-off-by: Bjarki Arge Andreasen <baa@trackunit.com>

Commit:
drivers/modem/modem_cmd_handler: Update API

There is currently not a clear separation between
user configuration and internal context when using the
modem_cmd_handler library. This update adds a clear
separation, placing user configuration in a seperate
struct passed to modem_cmd_handler_init alongside the
internal context modem_cmd_handler_data.

There is also a lack of documentation of the user
configurations, these have been added to the new config
struct.

The new API function modem_cmd_handler_process has been
added to remove the need for the user to directly access
the process member of the internal context. This ensures
that the user is not encouraged to access any internal
context members.

Some whitespace errors exist in the modem_cmd_handler.c
file, these are outside of the scope of this PR. These
can be addressed in a later PR as they are not functional
changes.

Signed-off-by: Bjarki Arge Andreasen <baa@trackunit.com>

Commit:
drivers/modem/modem_socket: Updated API

The internal socket context struct modem_socket_config
currently has members accessed directly by user. The
modem_socket_init() function has been updated to take all
user configurations as args. Thus removing the need for
the user to directly access the internal context for
initialization.

The user also currently needs to know of internal modem
socket libary behavior to determine if a socket has been
allocated and assigned an id, this is documented, and is
not safe. The functions:
 modem_socket_is_allocated()
 modem_socket_id_is_assigned()
 modem_socket_id_assign()
have been added to the modem socket library API to perform
these checks, and to assign socket ids.

This commit makes use of the modem socket library safer and
adds documentation to the API.

Signed-off-by: Bjarki Arge Andreasen <baa@trackunit.com>

bjarki-trackunit commented:
@mniestroj Every commit now contains motivation for each change, and every commit is bisectable. The three commits update modem_cmd_handler, modem_iface_uart and modem_socket seperately. The commit descriptions should answer most of the questions you raised.

@ycsin All additions to the modem helper libraries and driver modifications should now be formatted correctly, formatting of existing code has not been altered. Fixing formatting of existing code is outside the scope of this PR.

ycsin commented on drivers/modem/modem_cmd_handler.h:
```diff
@@ -266,15 +266,54 @@ int modem_cmd_handler_setup_cmds_nolock(struct modem_iface *iface,
+ * @param unsol_cmds_len Length of unsolicitet command handlers array
+ */
+struct modem_cmd_handler_setup {
+	char *match_buf;
+	size_t match_buf_len;
+	struct net_buf_pool *buf_pool;
+	k_timeout_t alloc_timeout;
+	const char *eol;
+	void *user_data;
+	const struct modem_cmd *response_cmds;
+	size_t response_cmds_len;
+	const struct modem_cmd *unsol_cmds;
+	size_t unsol_cmds_len;
+};
```
@bjarki-trackunit not sure if you can do something about this in this PR?

rerickson1 commented:
@mniestroj ping for re-review, thanks!

rerickson1 commented:
@mniestroj can you please review?
